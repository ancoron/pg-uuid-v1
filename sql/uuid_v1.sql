SET timezone TO 'Zulu';
\x

-- I/O
SELECT 'edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid_v1 AS ext;
SELECT 'EDB4D8F0-1A80-11E8-98D9-E03F49F7F8F3'::uuid_v1 AS ext_upper;
SELECT '{edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3}'::uuid_v1 AS ext_braces;
SELECT 'edb4d8f01a8011e898d9e03f49f7f8f3'::uuid_v1 AS ext_nosep;
SELECT '{edb4d8f01a8011e898d9e03f49f7f8f3}'::uuid_v1 AS ext_nosep_braces;
SELECT '.~([ edb4ZRd8f0P1a80_11e8/98d9 e03f49___f7f8f3 ])~.'::uuid_v1 AS ext_skip_garbage;

-- ...don't accept garbage...
SELECT 'edb4d8f0-1a80-11e8-98d9-e03_49f7f8f3'::uuid_v1 AS fail;

-- ...don't accept different versions...
SELECT '87c771ce-bc95-3114-ae59-c0e26acf8e81'::uuid_v1 AS ver_3;
SELECT '22859369-3a4f-49ef-8264-1aaf0a953299'::uuid_v1 AS ver_4;
SELECT 'c9aec822-6992-5c93-b34a-33cc0e952b5e'::uuid_v1 AS ver_5;
SELECT '00000000-0000-0000-0000-000000000000'::uuid_v1 AS nil;

-- ...don't accept different variants...
SELECT 'edb4d8f0-1a80-11e8-78d9-e03f49f7f8f3'::uuid_v1 AS var_ncs;
SELECT 'edb4d8f0-1a80-11e8-d8d9-e03f49f7f8f3'::uuid_v1 AS var_ms;
SELECT 'edb4d8f0-1a80-11e8-f8d9-e03f49f7f8f3'::uuid_v1 AS var_future;

-- conversion
SELECT pg_typeof(uuid_v1_convert('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid)) AS from_std;
SELECT pg_typeof(uuid_v1_convert('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid_v1)) AS to_std;


-- extract information
SELECT extract(epoch from ts), to_char(ts, 'YYYY-MM-DD HH24:MI:SS.US')
FROM uuid_v1_get_timestamp('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid_v1) AS ts;

SELECT extract(epoch from ts), to_char(ts, 'YYYY-MM-DD HH24:MI:SS.US')
FROM uuid_v1_get_timestamp('4938f30e-8449-11e9-ae2b-e03f49467033'::uuid_v1) AS ts;

SELECT uuid_v1_get_clockseq('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid_v1) AS clock_seq;
SELECT uuid_v1_get_clockseq('4938f30e-8449-11e9-ae2b-e03f49467033'::uuid_v1) AS clock_seq;

SELECT uuid_v1_get_node('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid_v1) AS node;
SELECT uuid_v1_get_node('4938f30e-8449-11e9-ae2b-e03f49467033'::uuid_v1) AS node;

-- basic comparison method
SELECT
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS eq,
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded3-8dbb-11e9-ae2b-db6f0f573554') AS lt,
    uuid_v1_cmp('8385ded3-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt,
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2c-db6f0f573554') AS lt_clock,
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2c-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt_clock,
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573555') AS lt_node,
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573555', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt_node
;

-- simple data tests
CREATE TABLE uuid_v1_test (id uuid_v1 PRIMARY KEY);

INSERT INTO uuid_v1_test (id) VALUES
('1004cd50-4241-11e9-b3ab-db6f0f573554'), -- 2019-03-09 07:58:02.056840
('05602550-8a8c-11e9-b3ab-db6f0f573554'), -- 2019-06-09 07:56:00.175240
('8385ded2-8dbb-11e9-ae2b-db6f0f573554'), -- 2019-06-13 09:13:31.650017
('ffc449f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-8f34-e03f49c7763b'), -- 2019-06-11 10:02:19.391640
('ffced5f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.460760
('ffd961f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.529880
('ffe3edf0-8c2f-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.599000
('ffee79f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.668120
('fff905f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.737240
('000391f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.806360
('000e1df0-8c30-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.875480
('0018a9f0-8c30-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.944600
('002335f0-8c30-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:20.013720
('002dc1f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:20.082840
('00384df0-8c30-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:20.151960
('0042d9f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:20.221080
('004d65f0-8c30-11e9-9bb8-e03f4977f7b7')  -- 2019-06-11 10:02:20.290200
;

ANALYZE uuid_v1_test;

SELECT
    count(*) FILTER (WHERE id < '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_lt,
    count(*) FILTER (WHERE id <= '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_le,
    count(*) FILTER (WHERE id > '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_gt,
    count(*) FILTER (WHERE id >= '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_ge
FROM uuid_v1_test;

SELECT
    count(*) FILTER (WHERE id <~ '2019-06-11 10:02:20.013720') AS count_lt,
    count(*) FILTER (WHERE id <=~ '2019-06-11 10:02:20.013720') AS count_le,
    count(*) FILTER (WHERE id >~ '2019-06-11 10:02:20.013720') AS count_gt,
    count(*) FILTER (WHERE id >=~ '2019-06-11 10:02:20.013720') AS count_ge
FROM uuid_v1_test;

-- verify use of sort-support
SET enable_seqscan TO off;
\x
SET timezone TO 'Asia/Tokyo';
EXPLAIN (ANALYZE, TIMING OFF, SUMMARY OFF, COSTS OFF)
SELECT count(*) FROM uuid_v1_test WHERE id <~ '2019-06-11 10:02:19Z';

EXPLAIN (ANALYZE, TIMING OFF, SUMMARY OFF, COSTS OFF)
SELECT * FROM uuid_v1_test WHERE id = '000e1df0-8c30-11e9-9bb8-e03f4977f7b7';

EXPLAIN (ANALYZE, TIMING OFF, SUMMARY OFF, COSTS OFF)
SELECT id FROM uuid_v1_test WHERE id < '002335f0-8c30-11e9-9bb8-e03f4977f7b7' ORDER BY id LIMIT 3 OFFSET 1;
