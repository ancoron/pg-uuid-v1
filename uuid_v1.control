# Special data type for version 1 UUID's
comment = 'UUID version 1'
default_version = '0.1'
module_pathname = '$libdir/uuid_v1'
relocatable = true
