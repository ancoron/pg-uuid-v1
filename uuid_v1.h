/*-------------------------------------------------------------------------
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 *
 * uuid_v1.h
 *	  Header file for the "uuid_v1" ADT. In C, we use the name pg_uuid_v1,
 *	  to avoid conflicts with any uuid_t type that might be defined by
 *	  the system headers.
 *
 * uuid_v1.h
 *
 *-------------------------------------------------------------------------
 */
#ifndef UUID_V1_H
#define UUID_V1_H

#define UUID_NODE_LEN 6

typedef struct pg_uuid_v1
{
    int64           timestamp;
    int16           clock_seq;
    unsigned char   node[UUID_NODE_LEN];
} pg_uuid_v1;

/* fmgr interface macros */
#define UUIDV1PGetDatum(X)		PointerGetDatum(X)
#define PG_RETURN_UUIDV1_P(X)	return UUIDV1PGetDatum(X)
#define DatumGetUUIDV1P(X)		((pg_uuid_v1 *) DatumGetPointer(X))
#define PG_GETARG_UUIDV1_P(X)	DatumGetUUIDV1P(PG_GETARG_DATUM(X))

extern pg_uuid_v1* uuid_std_to_v1(const pg_uuid_t *uuid);
extern pg_uuid_t* uuid_v1_to_std(const pg_uuid_v1 *uuid);
extern TimestampTz uuid_v1_timestamptz(const pg_uuid_v1 *uuid);

#endif							/* UUID_V1_H */
