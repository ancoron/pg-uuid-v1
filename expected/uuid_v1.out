SET timezone TO 'Zulu';
\x
-- I/O
SELECT 'edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid_v1 AS ext;
-[ RECORD 1 ]-----------------------------
ext | edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3

SELECT 'EDB4D8F0-1A80-11E8-98D9-E03F49F7F8F3'::uuid_v1 AS ext_upper;
-[ RECORD 1 ]-----------------------------------
ext_upper | edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3

SELECT '{edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3}'::uuid_v1 AS ext_braces;
-[ RECORD 1 ]------------------------------------
ext_braces | edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3

SELECT 'edb4d8f01a8011e898d9e03f49f7f8f3'::uuid_v1 AS ext_nosep;
-[ RECORD 1 ]-----------------------------------
ext_nosep | edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3

SELECT '{edb4d8f01a8011e898d9e03f49f7f8f3}'::uuid_v1 AS ext_nosep_braces;
-[ RECORD 1 ]----+-------------------------------------
ext_nosep_braces | edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3

SELECT '.~([ edb4ZRd8f0P1a80_11e8/98d9 e03f49___f7f8f3 ])~.'::uuid_v1 AS ext_skip_garbage;
-[ RECORD 1 ]----+-------------------------------------
ext_skip_garbage | edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3

-- ...don't accept garbage...
SELECT 'edb4d8f0-1a80-11e8-98d9-e03_49f7f8f3'::uuid_v1 AS fail;
ERROR:  invalid input syntax for type uuid_v1: "edb4d8f0-1a80-11e8-98d9-e03_49f7f8f3"
LINE 1: SELECT 'edb4d8f0-1a80-11e8-98d9-e03_49f7f8f3'::uuid_v1 AS fa...
               ^
-- ...don't accept different versions...
SELECT '87c771ce-bc95-3114-ae59-c0e26acf8e81'::uuid_v1 AS ver_3;
ERROR:  invalid version for type uuid_v1: "87c771ce-bc95-3114-ae59-c0e26acf8e81"
LINE 1: SELECT '87c771ce-bc95-3114-ae59-c0e26acf8e81'::uuid_v1 AS ve...
               ^
SELECT '22859369-3a4f-49ef-8264-1aaf0a953299'::uuid_v1 AS ver_4;
ERROR:  invalid version for type uuid_v1: "22859369-3a4f-49ef-8264-1aaf0a953299"
LINE 1: SELECT '22859369-3a4f-49ef-8264-1aaf0a953299'::uuid_v1 AS ve...
               ^
SELECT 'c9aec822-6992-5c93-b34a-33cc0e952b5e'::uuid_v1 AS ver_5;
ERROR:  invalid version for type uuid_v1: "c9aec822-6992-5c93-b34a-33cc0e952b5e"
LINE 1: SELECT 'c9aec822-6992-5c93-b34a-33cc0e952b5e'::uuid_v1 AS ve...
               ^
SELECT '00000000-0000-0000-0000-000000000000'::uuid_v1 AS nil;
ERROR:  invalid version for type uuid_v1: "00000000-0000-0000-0000-000000000000"
LINE 1: SELECT '00000000-0000-0000-0000-000000000000'::uuid_v1 AS ni...
               ^
-- ...don't accept different variants...
SELECT 'edb4d8f0-1a80-11e8-78d9-e03f49f7f8f3'::uuid_v1 AS var_ncs;
ERROR:  invalid variant for type uuid_v1: "edb4d8f0-1a80-11e8-78d9-e03f49f7f8f3"
LINE 1: SELECT 'edb4d8f0-1a80-11e8-78d9-e03f49f7f8f3'::uuid_v1 AS va...
               ^
SELECT 'edb4d8f0-1a80-11e8-d8d9-e03f49f7f8f3'::uuid_v1 AS var_ms;
ERROR:  invalid variant for type uuid_v1: "edb4d8f0-1a80-11e8-d8d9-e03f49f7f8f3"
LINE 1: SELECT 'edb4d8f0-1a80-11e8-d8d9-e03f49f7f8f3'::uuid_v1 AS va...
               ^
SELECT 'edb4d8f0-1a80-11e8-f8d9-e03f49f7f8f3'::uuid_v1 AS var_future;
ERROR:  invalid variant for type uuid_v1: "edb4d8f0-1a80-11e8-f8d9-e03f49f7f8f3"
LINE 1: SELECT 'edb4d8f0-1a80-11e8-f8d9-e03f49f7f8f3'::uuid_v1 AS va...
               ^
-- conversion
SELECT pg_typeof(uuid_v1_convert('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid)) AS from_std;
-[ RECORD 1 ]-----
from_std | uuid_v1

SELECT pg_typeof(uuid_v1_convert('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid_v1)) AS to_std;
-[ RECORD 1 ]
to_std | uuid

-- extract information
SELECT extract(epoch from ts), to_char(ts, 'YYYY-MM-DD HH24:MI:SS.US')
FROM uuid_v1_get_timestamp('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid_v1) AS ts;
-[ RECORD 1 ]-------------------------
date_part | 1519600168.09884
to_char   | 2018-02-25 23:09:28.098840

SELECT extract(epoch from ts), to_char(ts, 'YYYY-MM-DD HH24:MI:SS.US')
FROM uuid_v1_get_timestamp('4938f30e-8449-11e9-ae2b-e03f49467033'::uuid_v1) AS ts;
-[ RECORD 1 ]-------------------------
date_part | 1559378590.74588
to_char   | 2019-06-01 08:43:10.745883

SELECT uuid_v1_get_clockseq('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid_v1) AS clock_seq;
-[ RECORD 1 ]---
clock_seq | 6361

SELECT uuid_v1_get_clockseq('4938f30e-8449-11e9-ae2b-e03f49467033'::uuid_v1) AS clock_seq;
-[ RECORD 1 ]----
clock_seq | 11819

SELECT uuid_v1_get_node('edb4d8f0-1a80-11e8-98d9-e03f49f7f8f3'::uuid_v1) AS node;
-[ RECORD 1 ]--------
node | \xe03f49f7f8f3

SELECT uuid_v1_get_node('4938f30e-8449-11e9-ae2b-e03f49467033'::uuid_v1) AS node;
-[ RECORD 1 ]--------
node | \xe03f49467033

-- basic comparison method
SELECT
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS eq,
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded3-8dbb-11e9-ae2b-db6f0f573554') AS lt,
    uuid_v1_cmp('8385ded3-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt,
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2c-db6f0f573554') AS lt_clock,
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2c-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt_clock,
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573554', '8385ded2-8dbb-11e9-ae2b-db6f0f573555') AS lt_node,
    uuid_v1_cmp('8385ded2-8dbb-11e9-ae2b-db6f0f573555', '8385ded2-8dbb-11e9-ae2b-db6f0f573554') AS gt_node
;
-[ RECORD 1 ]
eq       | 0
lt       | -1
gt       | 1
lt_clock | -1
gt_clock | 1
lt_node  | -1
gt_node  | 1

-- simple data tests
CREATE TABLE uuid_v1_test (id uuid_v1 PRIMARY KEY);
INSERT INTO uuid_v1_test (id) VALUES
('1004cd50-4241-11e9-b3ab-db6f0f573554'), -- 2019-03-09 07:58:02.056840
('05602550-8a8c-11e9-b3ab-db6f0f573554'), -- 2019-06-09 07:56:00.175240
('8385ded2-8dbb-11e9-ae2b-db6f0f573554'), -- 2019-06-13 09:13:31.650017
('ffc449f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.391640
('ffc449f0-8c2f-11e9-8f34-e03f49c7763b'), -- 2019-06-11 10:02:19.391640
('ffced5f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.460760
('ffd961f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.529880
('ffe3edf0-8c2f-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.599000
('ffee79f0-8c2f-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.668120
('fff905f0-8c2f-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.737240
('000391f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:19.806360
('000e1df0-8c30-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:19.875480
('0018a9f0-8c30-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:19.944600
('002335f0-8c30-11e9-9bb8-e03f4977f7b7'), -- 2019-06-11 10:02:20.013720
('002dc1f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:20.082840
('00384df0-8c30-11e9-96b4-e03f49d7f7bb'), -- 2019-06-11 10:02:20.151960
('0042d9f0-8c30-11e9-aba7-e03f497ffcbf'), -- 2019-06-11 10:02:20.221080
('004d65f0-8c30-11e9-9bb8-e03f4977f7b7')  -- 2019-06-11 10:02:20.290200
;
ANALYZE uuid_v1_test;
SELECT
    count(*) FILTER (WHERE id < '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_lt,
    count(*) FILTER (WHERE id <= '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_le,
    count(*) FILTER (WHERE id > '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_gt,
    count(*) FILTER (WHERE id >= '002335f0-8c30-11e9-9bb8-e03f4977f7b7') AS count_ge
FROM uuid_v1_test;
-[ RECORD 1 ]
count_lt | 14
count_le | 15
count_gt | 5
count_ge | 6

SELECT
    count(*) FILTER (WHERE id <~ '2019-06-11 10:02:20.013720') AS count_lt,
    count(*) FILTER (WHERE id <=~ '2019-06-11 10:02:20.013720') AS count_le,
    count(*) FILTER (WHERE id >~ '2019-06-11 10:02:20.013720') AS count_gt,
    count(*) FILTER (WHERE id >=~ '2019-06-11 10:02:20.013720') AS count_ge
FROM uuid_v1_test;
-[ RECORD 1 ]
count_lt | 14
count_le | 15
count_gt | 5
count_ge | 6

-- verify use of sort-support
SET enable_seqscan TO off;
\x
SET timezone TO 'Asia/Tokyo';
EXPLAIN (ANALYZE, TIMING OFF, SUMMARY OFF, COSTS OFF)
SELECT count(*) FROM uuid_v1_test WHERE id <~ '2019-06-11 10:02:19Z';
                                      QUERY PLAN                                       
---------------------------------------------------------------------------------------
 Aggregate (actual rows=1 loops=1)
   ->  Index Only Scan using uuid_v1_test_pkey on uuid_v1_test (actual rows=2 loops=1)
         Index Cond: (id <~ 'Tue Jun 11 19:02:19 2019 JST'::timestamp with time zone)
         Heap Fetches: 2
(4 rows)

EXPLAIN (ANALYZE, TIMING OFF, SUMMARY OFF, COSTS OFF)
SELECT * FROM uuid_v1_test WHERE id = '000e1df0-8c30-11e9-9bb8-e03f4977f7b7';
                                   QUERY PLAN                                    
---------------------------------------------------------------------------------
 Index Only Scan using uuid_v1_test_pkey on uuid_v1_test (actual rows=1 loops=1)
   Index Cond: (id = '000e1df0-8c30-11e9-9bb8-e03f4977f7b7'::uuid_v1)
   Heap Fetches: 1
(3 rows)

EXPLAIN (ANALYZE, TIMING OFF, SUMMARY OFF, COSTS OFF)
SELECT id FROM uuid_v1_test WHERE id < '002335f0-8c30-11e9-9bb8-e03f4977f7b7' ORDER BY id LIMIT 3 OFFSET 1;
                                      QUERY PLAN                                       
---------------------------------------------------------------------------------------
 Limit (actual rows=3 loops=1)
   ->  Index Only Scan using uuid_v1_test_pkey on uuid_v1_test (actual rows=4 loops=1)
         Index Cond: (id < '002335f0-8c30-11e9-9bb8-e03f4977f7b7'::uuid_v1)
         Heap Fetches: 4
(4 rows)

